<a href="https://colab.research.google.com/github/drshahizan/Python-big-data/blob/main/assignment/ass6/bdm/HR/big_data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Assignment 6: Mastering Big Data Handling**
##Dataset -- Newyork Taxi Trip Data

> Group Members:

 | Name          | Matric Number  |
 | ------------- | -------------- |
 | MOHAMMED RAZA ASFAK CHIDIMAR     | MCS231004       |
 | HUSSEIN YUSUF SHEIKH MOHAMED   | MCS231024       |

Dive into the rich dataset of New York City's yellow taxis, a treasure trove for any data scientist. This dataset is a detailed chronicle of urban mobility, capturing intricate details of each taxi trip. From temporal data points like pickup and drop-off times to spatial dimensions involving locations, this data offers a granular view of urban transit patterns. It also includes quantitative metrics like trip distance, fare composition, payment methods, and passenger counts. Provided by tech companies under the TPEP/LPEP initiatives, this data is a goldmine for analyzing urban transportation trends, fare dynamics, and passenger behavior. It's not just about taxis and their routes; it's a window into the lifeblood of the city, offering endless possibilities for data-driven insights and urban planning.


# Loading Dataset

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
! pip install kaggle
! mkdir ~/.kaggle

In [3]:
!cp /content/drive/MyDrive/kaggle.json ~/.kaggle/kaggle.json
! chmod 600 ~/.kaggle/kaggle.json

In [3]:
! kaggle datasets download -d microize/newyork-yellow-taxi-trip-data-2020-2019


100% 1.80G/1.81G [00:19<00:00, 155MB/s]
100% 1.81G/1.81G [00:19<00:00, 99.9MB/s]


In [4]:
! unzip newyork-yellow-taxi-trip-data-2020-2019

Archive:  newyork-yellow-taxi-trip-data-2020-2019.zip
  inflating: taxi+_zone_lookup.csv   
  inflating: taxi_zones/taxi_zones.dbf  
  inflating: taxi_zones/taxi_zones.prj  
  inflating: taxi_zones/taxi_zones.sbn  
  inflating: taxi_zones/taxi_zones.sbx  
  inflating: taxi_zones/taxi_zones.shp  
  inflating: taxi_zones/taxi_zones.shp.xml  
  inflating: taxi_zones/taxi_zones.shx  
  inflating: yellow_tripdata_2019-01.csv  
  inflating: yellow_tripdata_2019-02.csv  
  inflating: yellow_tripdata_2019-03.csv  
  inflating: yellow_tripdata_2019-04.csv  
  inflating: yellow_tripdata_2019-05.csv  
  inflating: yellow_tripdata_2019-06.csv  
  inflating: yellow_tripdata_2019-07.csv  
  inflating: yellow_tripdata_2019-08.csv  
  inflating: yellow_tripdata_2019-09.csv  
  inflating: yellow_tripdata_2019-10.csv  
  inflating: yellow_tripdata_2019-11.csv  
  inflating: yellow_tripdata_2019-12.csv  
  inflating: yellow_tripdata_2020-01.csv  
  inflating: yellow_tripdata_2020-02.csv  
  inflating: ye

In [5]:
import pandas as pd
import numpy as np

In [6]:
df = pd.read_csv("/content/yellow_tripdata_2019-03.csv")
df

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0               1  2019-03-01 00:24:41   2019-03-01 00:25:31                1   
1               1  2019-03-01 00:25:27   2019-03-01 00:36:37                2   
2               1  2019-03-01 00:05:21   2019-03-01 00:38:23                1   
3               1  2019-03-01 00:48:55   2019-03-01 01:06:03                1   
4               1  2019-03-01 00:11:42   2019-03-01 00:16:40                1   
...           ...                  ...                   ...              ...   
7832540         2  2019-03-31 23:58:52   2019-04-01 00:05:04                1   
7832541         2  2019-03-31 23:20:15   2019-03-31 23:24:12                1   
7832542         2  2019-03-31 23:32:15   2019-03-31 23:54:36                1   
7832543         2  2019-03-31 22:32:57   2019-03-31 23:21:22                1   
7832544         2  2019-03-31 23:31:13   2019-03-31 23:31:15                1   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
0                 0.00           1                  N           145   
1                 3.70           1                  N            95   
2                14.10           1                  N           249   
3                 9.60           1                  N           138   
4                 0.80           1                  N            48   
...                ...         ...                ...           ...   
7832540           0.74           1                  N           186   
7832541           1.11           1                  N           263   
7832542          13.17           1                  N            75   
7832543           0.00           1                  N           237   
7832544           0.00           1                  N           264   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
0                 145             2          2.5    0.5      0.5        0.00   
1                 130             1         13.0    0.5      0.5        0.70   
2                  28             1         41.0    3.0      0.5       10.10   
3                  98             2         27.0    0.5      0.5        0.00   
4                  48             1          5.5    3.0      0.5        3.00   
...               ...           ...          ...    ...      ...         ...   
7832540           186             1          5.5    0.5      0.5        2.33   
7832541           236             2          5.5    0.5      0.5        0.00   
7832542           134             2         37.0    0.5      0.5        0.00   
7832543           193             2          0.0    0.0      0.0        0.00   
7832544           264             2          0.0    0.0      0.0        0.00   

         tolls_amount  improvement_surcharge  total_amount  \
0                0.00                    0.3          3.80   
1                0.00                    0.3         15.00   
2                5.76                    0.3         60.66   
3                0.00                    0.3         28.30   
4                0.00                    0.3         12.30   
...               ...                    ...           ...   
7832540          0.00                    0.3         11.63   
7832541          0.00                    0.3          9.30   
7832542          5.76                    0.3         44.06   
7832543          0.00                    0.0          0.00   
7832544          0.00                    0.0          0.00   

         congestion_surcharge  
0                         0.0  
1                         0.0  
2                         2.5  
3                         0.0  
4                         2.5  
...                       ...  
7832540                   2.5  
7832541                   2.5  
7832542                   0.0  
7832543                   2.5  
7832544                   0.0  

[7832545 rows x 18 columns]

**Checking how much of memory is used when loading the entire CSV**

In [7]:
# Load and display information for the full CSV dataset
full_data_path = "/content/yellow_tripdata_2019-03.csv"
df = pd.read_csv(full_data_path)

# Display concise information about the DataFrame
df.info(verbose=False, memory_usage="deep")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7832545 entries, 0 to 7832544
Columns: 18 entries, VendorID to congestion_surcharge
dtypes: float64(9), int64(6), object(3)
memory usage: 2.4 GB


# Strategies for Big Datasets

In [8]:
# Importing necessary libraries
import matplotlib.pyplot as plt
import seaborn as sns
import random
import time
import psutil

### Load Less Data

In [9]:
# Display column names in the DataFrame
column_names = df.columns
column_names

Index(['VendorID', 'tpep_pickup_datetime', 'tpep_dropoff_datetime',
       'passenger_count', 'trip_distance', 'RatecodeID', 'store_and_fwd_flag',
       'PULocationID', 'DOLocationID', 'payment_type', 'fare_amount', 'extra',
       'mta_tax', 'tip_amount', 'tolls_amount', 'improvement_surcharge',
       'total_amount', 'congestion_surcharge'],
      dtype='object')

In [10]:
# Display DataFrame information with memory usage details
df.info(verbose=False, memory_usage="deep")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7832545 entries, 0 to 7832544
Columns: 18 entries, VendorID to congestion_surcharge
dtypes: float64(9), int64(6), object(3)
memory usage: 2.4 GB


In [11]:
columns_to_load = ['VendorID', 'tpep_pickup_datetime', 'tpep_dropoff_datetime', 'passenger_count', 'trip_distance', 'payment_type', 'fare_amount', 'total_amount']
df = pd.read_csv("/content/yellow_tripdata_2019-03.csv", usecols=columns_to_load)


By loading only the necessary data, memory usage decreased from 2.0 GB to 1015.60 MB. When the complete dataset is not required, loading only a portion of it is helpful for quicker analysis.

### Use Chunking

In [12]:
# Number of data of the DataFrame
len(df)

7832545

In [13]:
import time
import pandas as pd

# Record start time
start_time = time.time()

# Load the entire dataset
df = pd.read_csv("/content/yellow_tripdata_2019-03.csv")

# Record end time
end_time = time.time()

# Calculate and print the loading time
loading_time = end_time - start_time
print(f"Time to load the entire dataset: {loading_time:.2f} seconds")

Time to load the entire dataset: 38.33 seconds


Let's divide it into different chunks, each holding 100,000 entries of data.

In [14]:
import gc

chunk_size = 100000
df_chunks = pd.read_csv("/content/yellow_tripdata_2019-03.csv", chunksize=chunk_size)

num = 0
for chunk in df_chunks:
    num += 1
    chunk.to_csv(f"chunk{num}.csv", index=False)
    gc.collect()

num

79

In [15]:
random_chunk_number = random.randint(1, num)

# Store Start time
start_time_reading = time.time()

# Read the CSV chunk file
chunk_filename = f"chunk{random_chunk_number}.csv"
loaded_chunk = pd.read_csv(chunk_filename)

# Store End time
end_time_reading = time.time()

# Calculate reading time
reading_time = end_time_reading - start_time_reading
print(f"Reading Time: {reading_time:.2f} seconds")

# CPU and memory usage information
cpu_percent_usage = psutil.cpu_percent(interval=1)
memory_info = psutil.virtual_memory()

print(f"CPU Usage: {cpu_percent_usage}%")
print(f"Memory Usage: {memory_info.percent}%")

# Check system RAM usage
ram_percent_usage = psutil.virtual_memory().percent
print(f"System RAM Usage: {ram_percent_usage}%")

# Display the first few rows of the loaded chunk
loaded_chunk.head()

Reading Time: 0.54 seconds
CPU Usage: 81.6%
Memory Usage: 39.6%
System RAM Usage: 39.6%


VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         2  2019-03-15 16:15:30   2019-03-15 16:21:29                2   
1         2  2019-03-15 16:03:15   2019-03-15 16:59:44                1   
2         1  2019-03-15 16:22:29   2019-03-15 16:29:13                2   
3         1  2019-03-15 16:05:04   2019-03-15 17:23:40                1   
4         2  2019-03-15 15:58:09   2019-03-15 16:07:31                1   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0           1.38           1                  N           141           263   
1          17.39           2                  N           132            74   
2           1.00           1                  N           166           152   
3          19.70           2                  N           237           132   
4           0.71           1                  N           107           170   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0             1          6.5    1.0      0.5        2.16          0.00   
1             1         52.0    4.5      0.5       12.61          5.76   
2             1          6.5    1.0      0.5        2.05          0.00   
3             1         52.0    7.0      0.5       11.00          5.76   
4             1          7.0    1.0      0.5        2.26          0.00   

   improvement_surcharge  total_amount  congestion_surcharge  
0                    0.3         12.96                   2.5  
1                    0.3         75.67                   0.0  
2                    0.3         10.35                   0.0  
3                    0.3         76.56                   2.5  
4                    0.3         13.56                   2.5

So, loading each of the chunk datasets takes only 0.54s, compared to loading the entire dataset a once.

### Optimize Data Types

In [16]:
#Dataframe size before optimization

from sys import getsizeof
# The size of the original data set
start_size = getsizeof(df)/(1024.0**3)
print('Dataframe size: %2.2f GB'%start_size)

Dataframe size: 2.41 GB


In [17]:
df.dtypes

VendorID                   int64
tpep_pickup_datetime      object
tpep_dropoff_datetime     object
passenger_count            int64
trip_distance            float64
RatecodeID                 int64
store_and_fwd_flag        object
PULocationID               int64
DOLocationID               int64
payment_type               int64
fare_amount              float64
extra                    float64
mta_tax                  float64
tip_amount               float64
tolls_amount             float64
improvement_surcharge    float64
total_amount             float64
congestion_surcharge     float64
dtype: object

In [18]:
numeric_columns = df.select_dtypes(include=['int64', 'float64']).columns

min_values = df[numeric_columns].min()
max_values = df[numeric_columns].max()

print("Minimum values:\n", min_values)
print("\nMaximum values:\n", max_values)

Minimum values:
 VendorID                   1.00
passenger_count            0.00
trip_distance              0.00
RatecodeID                 1.00
PULocationID               1.00
DOLocationID               1.00
payment_type               1.00
fare_amount             -447.00
extra                    -57.50
mta_tax                   -0.50
tip_amount               -89.89
tolls_amount             -25.00
improvement_surcharge     -0.30
total_amount            -450.30
congestion_surcharge      -2.50
dtype: float64

Maximum values:
 VendorID                       4.00
passenger_count                9.00
trip_distance                237.17
RatecodeID                    99.00
PULocationID                 265.00
DOLocationID                 265.00
payment_type                   5.00
fare_amount               943274.80
extra                         24.66
mta_tax                       75.00
tip_amount                141492.02
tolls_amount                 805.76
improvement_surcharge          1.00
to

In [19]:
data_types_mapping = {
    'float64': 'float16',
    'int64': 'int16',
    'object': 'category'
}

df = df.apply(lambda col: col.astype(data_types_mapping.get(str(col.dtype), col.dtype)))

In [20]:
start_memory = df.memory_usage().sum() / 1024**3

# Optimize memory usage
df_optimized = df.copy()
for col in df_optimized.columns:
    if pd.api.types.is_numeric_dtype(df_optimized[col]):
        df_optimized[col] = pd.to_numeric(df_optimized[col], downcast='integer')
    elif pd.api.types.is_object_dtype(df_optimized[col]):
        df_optimized[col] = df_optimized[col].astype('category')

end_memory = df_optimized.memory_usage().sum() / 1024**3

# Report memory usage changes
print(f"Memory usage after optimization: {end_memory:.2f} MB")
print(f"Decreased by: {100 * (start_memory - end_memory) / start_memory:.1f}%")

Memory usage after optimization: 0.42 MB
Decreased by: 6.6%


In [21]:
# Display concise information about the DataFrame
summary_info = df.memory_usage(deep=True).sum() / (1024**3), df.shape
print(f"Total memory usage: {summary_info[0]:.2f} GB")
print("Data types:")
print(df.dtypes)
print(summary_info[1])

Total memory usage: 0.73 GB
Data types:
VendorID                    int16
tpep_pickup_datetime     category
tpep_dropoff_datetime    category
passenger_count             int16
trip_distance             float16
RatecodeID                  int16
store_and_fwd_flag       category
PULocationID                int16
DOLocationID                int16
payment_type                int16
fare_amount               float16
extra                     float16
mta_tax                   float16
tip_amount                float16
tolls_amount              float16
improvement_surcharge     float16
total_amount              float16
congestion_surcharge      float16
dtype: object
(7832545, 18)


In [22]:
start_size = getsizeof(df)/(1024.0**3)
print('Dataframe size: %2.2f GB'%start_size)

Dataframe size: 0.73 GB


So, following the optimization process, the memory usage decreased from 2.41 GB to 0.73 GB, representing a 69.7% reduction.

### Sampling

In [23]:
file_path = "/content/yellow_tripdata_2019-03.csv"

# Read CSV file into a Pandas DataFrame
df = pd.read_csv(file_path)

# Check memory usage of the DataFrame
memory_usage = df.memory_usage(deep=True).sum() / (1024 * 1024)  # in megabytes

# Print the memory usage
print(f"Memory Usage: {memory_usage:.2f} MB")

Memory Usage: 2465.00 MB


In [24]:
# Retrieve the dimensions (rows and columns) of the DataFrame
dimensions = df.shape
dimensions

(7832545, 18)

In [25]:
# Randomly sample 10% of the dataset
sample_10_percent = df.sample(frac=0.10)
sample_10_percent

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
6153145         2  2019-03-25 06:26:01   2019-03-25 07:25:15                6   
4158781         1  2019-03-16 21:44:56   2019-03-16 21:54:02                1   
6963484         2  2019-03-28 14:57:17   2019-03-28 15:10:30                3   
7767666         2  2019-03-31 17:02:17   2019-03-31 17:04:42                1   
3016422         2  2019-03-12 18:11:04   2019-03-12 19:06:20                2   
...           ...                  ...                   ...              ...   
514358          2  2019-03-02 22:35:34   2019-03-02 22:41:13                1   
326523          2  2019-03-02 07:31:42   2019-03-02 07:35:52                1   
1729911         1  2019-03-07 19:40:37   2019-03-07 19:57:29                1   
1396438         2  2019-03-06 16:36:34   2019-03-07 16:30:28                4   
4555818         2  2019-03-18 16:09:14   2019-03-18 16:12:50                1   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
6153145          19.98           2                  N           132   
4158781           1.10           1                  N            90   
6963484           2.19           1                  N           158   
7767666           0.55           1                  N           140   
3016422          18.68           2                  N           132   
...                ...         ...                ...           ...   
514358            1.27           1                  N            90   
326523            0.58           1                  N           262   
1729911           2.20           1                  N           161   
1396438           6.52           1                  N           162   
4555818           0.53           1                  N           233   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
6153145           238             1         52.0    0.0      0.5       11.71   
4158781           114             2          7.5    3.0      0.5        0.00   
6963484            90             2         10.5    0.0      0.5        0.00   
7767666           229             1          4.0    0.0      0.5        1.46   
3016422            43             2         52.0    4.5      0.5        0.00   
...               ...           ...          ...    ...      ...         ...   
514358            246             1          6.5    0.5      0.5        2.58   
326523            236             1          5.0    0.0      0.5        2.08   
1729911           246             2         12.0    3.0      0.5        0.00   
1396438           231             1         24.5    1.0      0.5        0.00   
4555818           170             1          4.5    1.0      0.5        1.00   

         tolls_amount  improvement_surcharge  total_amount  \
6153145          5.76                    0.3         70.27   
4158781          0.00                    0.3         11.30   
6963484          0.00                    0.3         13.80   
7767666          0.00                    0.3          8.76   
3016422          5.76                    0.3         65.56   
...               ...                    ...           ...   
514358           0.00                    0.3         12.88   
326523           0.00                    0.3         10.38   
1729911          0.00                    0.3         15.80   
1396438          0.00                    0.3         28.80   
4555818          0.00                    0.3          9.80   

         congestion_surcharge  
6153145                   0.0  
4158781                   2.5  
6963484                   2.5  
7767666                   2.5  
3016422                   2.5  
...                       ...  
514358                    2.5  
326523                    2.5  
1729911                   2.5  
1396438                   2.5  
4555818                   2.5  

[783254 rows x 18 columns]

In [26]:
# Show memory usage details for the 10% sample
memory_info_10_percent = sample_10_percent.memory_usage(deep=True).sum() / (1024**2)
print(f"Memory usage after 10% sampling: {memory_info_10_percent:.2f} MB")

Memory usage after 10% sampling: 252.48 MB


Therefore, after sampling the memory usage significantly decreased from 2465.00 MB to 252.48 MB, resulting in approximately 89.76% reduction when using 10% of the sample.

### Parallelize with Dask

In this part, we'll use pandas and Dask methods to calculate various data points and draw a comparison between them.



*   Reading the Dataframe



In [27]:
# Import necessary libraries
import dask.dataframe as dd
import time

In [48]:
# Reading the dataframe with Pandas

start_time_pandas = time.time()

df3_pd = pd.read_csv("/content/yellow_tripdata_2019-03.csv")

end_time_pandas = time.time()
execution_time_pandas = end_time_pandas - start_time_pandas

print(f"Pandas Execution Time: {execution_time_pandas} seconds")

Pandas Execution Time: 43.496740102767944 seconds


In [53]:
# Measure time before
start_time = time.time()

# Reading a DataFrame with Dask
df3_dd = dd.read_csv("/content/yellow_tripdata_2019-03.csv")

# Measure time after
end_time = time.time()

execution_time_dask = end_time - start_time
print(f"Reading DataFrame with Dask Execution Time: {execution_time_dask} seconds")

Reading DataFrame with Dask Execution Time: 0.024603843688964844 seconds




*   Calculation time to find the Mean of the fare amount of VendorIDs


In [64]:
#Using Pandas
start_time_pandas = time.time()
result_pandas = df3_pd.groupby('VendorID')['fare_amount'].mean()
end_time_pandas = time.time()

# Display the results and computational time using Pandas
print("Mean of the fare amount of VendorIDs:\n", result_pandas)
print("Pandas Computational Time:", end_time_pandas - start_time_pandas, "seconds")

Mean of the fare amount of VendorIDs:
 VendorID
1    13.016585
2    13.285019
4    12.292509
Name: fare_amount, dtype: float64
Pandas Computational Time: 0.20585322380065918 seconds


In [85]:
#Using Dask
ddf = dd.from_pandas(df, npartitions=3)


In [86]:
start_time_dask = time.time()
result = ddf.groupby('VendorID')['fare_amount'].mean().compute()
end_time_dask = time.time()

# Display the results and computational time using Dask
print("Mean of the fare amount of VendorIDs:\n", result)
print("Dask Computational Time:", end_time_dask - start_time_dask, "seconds")

Mean of the fare amount of VendorIDs:
 VendorID
1    13.016585
2    13.285019
4    12.292509
Name: fare_amount, dtype: float64
Dask Computational Time: 0.5796234607696533 seconds


It took 43.50 seconds for Pandas to read the dataframe, but only 0.025 seconds for Dask. And calculating the Mean of the fare amount of VendorIDs with Pandas took only 0.21 seconds, whereas Dask took 0.58 seconds.

## Comparative Analysis

| No. | Traditional Method | Advanced Strategies | Result |
   | --- | ------------------ | -------------------- | ------ |
   | 1.  | Memory usage 2.00 GB | **Load Less Data**: Memory usage 1015.60 MB | Memory usage reduced by 49.20% |
   | 2.  | Time taken to read dataframe 42.99 s| **Use Chunking**: Time taken to read in chunks 0.67 s | Reading time reduced by 98.4% |
   | 3.  | Data Frame Size 2.00 GB | **Optimize Data Types**: After optimization 0.89 GB | Data frame size reduced by 55.50% |
   | 4.  | Memory usage 2051.85 MB | **Sampling**: 10% sampling; Memory usage 209.73 MB | Memory usage reduced by 89.8% |
   | 5.  | Reading Data frame: 43.48 s | **Parallelize with Dask**: Reading Data frame: 0.024 s | Processing time reading dataframe reduced by 99.94% |

### Pros & Cons

**1. Pandas**

Pros:


*   Powerful data structures and functions for analysis and manipulation
*   Efficient for smaller datasets
*   Intuitive syntax and rich ecosystem of tools

Cons:


*   Can consume significant memory for large datasets
*   Operations on large DataFrames can be slow

**2. Load Less Data:**


Pros:


*   Reduces memory footprint and initial loading time
*   Can improve performance for specific tasks


Cons:


*   Requires careful selection of relevant data
*   Could limit analysis if essential data is excluded





**3. Use Chunking**

Pros:


*   Processes large datasets in manageable chunks
*   Reduces memory overhead and can improve performance for certain operations.


Cons:


*   Requires careful handling of data continuity
*   Can introduce code complexity

**4. Optimize Data Types**

Pros:


*   Reduces memory usage by choosing appropriate data types
*   Can improve performance for computations

Cons:


*   Requires understanding of data types and their memory implications
*   Might necessitate data type conversions

**5. Sampling**

Pros:


*   Analyzes a representative subset to reduce memory and processing time
*   Can provide insights without processing the entire dataset


Cons:


*   Potential loss of information if sample is not representative
*   Sampling bias might lead to inaccurate conclusions



**6. Parallelize with Dask**

Pros:


*   Distributes computations across multiple cores or machines
*   Can handle datasets too large for single-machine processing

Cons:


*   Adds complexity to code and setup
*   Requires understanding of parallel computing principles
*   Might not be suitable for all tasks due to overhead

## Conclusion




In conclusion, the comparative analysis between traditional methods and advanced strategies demonstrates substantial improvements in memory usage, processing time, and data frame size.Firstly, memory usage has significantly decreased by 49.20% through the implementation of strategies such as loading less data and optimizing data types. Secondly, substantial improvements in reading time have been achieved, with a remarkable 98.4% reduction in the time taken to read data frames. Moreover, the optimization of data types has led to a noteworthy reduction of 55.50% in data frame size. This not only aids in better storage utilization but also facilitates faster data retrieval and manipulation. Lastly, the combination of sampling and parallelization with Dask has resulted in an impressive 89.8% decrease in memory usage during data frame reading. These advanced strategies collectively showcase the potential for substantial efficiency gains, making them essential considerations for anyone seeking to enhance the performance of data processing workflows.